src/core/net/netif.hpp

```diff
@@ -164,8 +200,125 @@ class Netif : public InstanceLocator, public LinkedListEntry<Netif>, private Non
+         * empty.
+         *
+         */
+        void operator++(int) { AdvanceFrom(mCurrent->GetNext()); }
+
+        /**
+         * This method overloads the `*` dereference operator and gets a reference to `ExternalNetifMulticastAddress`
+         * entry to which the iterator is currently pointing.
+         *
+         * This method MUST be used when the iterator is not empty.
+         *
+         * @returns A reference to the `ExternalNetifMulticastAddress` entry currently pointed by the iterator.
+         *
+         */
+        ExternalNetifMulticastAddress &operator*(void) { return *mCurrent; }
+
+        /**
+         * This method overloads the `->` dereference operator and gets a pointer to `ExternalNetifMulticastAddress`
+         * entry to which the iterator is current pointing.
+         *
+         * @returns A pointer to the `ExternalNetifMulticastAddress` entry associated with the iterator, or `nullptr` if
+         * iterator is empty.
+         *
+         */
+        ExternalNetifMulticastAddress *operator->(void) { return mCurrent; }
+
+        /**
+         * This method overloads operator `==` to evaluate whether or not two `ExternalMulticastAddressIterator`
+         * instances point to the same `ExternalNetifMulticastAddress` entry.
+         *
+         * @param[in] aOther  The other `Iterator` to compare with.
+         *
+         * @retval TRUE   If the two `ExternalMulticastAddressIterator` objects point to the same
+         * `ExternalNetifMulticastAddress` entry or both are done.
+         * @retval FALSE  If the two `ExternalMulticastAddressIterator` objects do not point to the same
+         * `ExternalNetifMulticastAddress` entry.
+         *
+         */
+        bool operator==(const ExternalMulticastAddressIterator &aOther) { return mCurrent == aOther.mCurrent; }
+
+        /**
+         * This method overloads operator `!=` to evaluate whether or not two `ExternalMulticastAddressIterator`
+         * instances point to the same `ExternalNetifMulticastAddress` entry.
+         *
+         * @param[in]  aOther  The other `ExternalMulticastAddressIterator` to compare with.
+         *
+         * @retval TRUE   If the two `ExternalMulticastAddressIterator` objects do not point to the same
+         * `ExternalNetifMulticastAddress` entry.
+         * @retval FALSE  If the two `ExternalMulticastAddressIterator` objects point to the same
+         * `ExternalNetifMulticastAddress` entry or both are done.
+         *
+         */
+        bool operator!=(const ExternalMulticastAddressIterator &aOther) { return mCurrent != aOther.mCurrent; }
+
+    private:
+        enum IteratorType
+        {
+            kEndIterator,
+        };
+
+        ExternalMulticastAddressIterator(Netif &aNetif, IteratorType)
+            : mNetif(aNetif)
+            , mCurrent(nullptr)
+        {
+        }
+
+        void AdvanceFrom(NetifMulticastAddress *aAddr)
```

Comment:
Can we consider defining `aAddr` as `const NetifMulticastAddress *`?

It would then help with `AdvanceFrom(const_cast<NetifMulticastAddress *>(mNetif.GetMulticastAddresses()));` (no need for the `const_cast`) and I think it should work ok with `AdvanceFrom(mCurrent->GetNext());` as well?

Reply:
revised.